### ISSUE TITLE
Try using batch statements again

### ISSUE THREAD

**Issue by patriknw** [akka/akka-persistence-r2dbc#87]
Something didn't work when I tried them, but maybe that was related to the concurrency problems we had.

Note that there are two types of batch statements.
1) Statements with bind parameters that are separated with `.add()`
2) Batch without bind parameters

I don't think 2 is interesting for us because we want use bind parameters (e.g. binary param).

**Comment by octonato** [comment]
It seems that this is a problem in the driver itself. 

First, if I run the tests and inspect the DB I see the following: 

**with batch inserts**

```
 persistence_id | sequence_number |         db_timestamp  
----------------+-----------------+-------------------------------
 p-1            |               1 | 2021-11-09 11:46:42.928996+00
 p-2            |               1 | 2021-11-09 11:46:43.141532+00
 p-3            |               1 | 2021-11-09 11:46:43.194956+00
 p-4            |               1 | 2021-11-09 11:46:43.262527+00
 p-5            |               1 | 2021-11-09 11:46:43.327334+00
 p-6            |               1 | 2021-11-09 11:46:43.378623+00
 p-7            |               1 | 2021-11-09 11:46:43.426329+00
 p-8            |               1 | 2021-11-09 11:46:43.476555+00
 p-9            |               1 | 2021-11-09 11:46:43.52645+00
 p-10           |               1 | 2021-11-09 11:46:43.571498+00
 p-11           |               1 | 2021-11-09 11:46:43.620198+00
 p-12           |               3 | 2021-11-09 11:46:43.706853+00
 p-13           |               1 | 2021-11-09 11:46:43.762194+00
 p-14           |               1 | 2021-11-09 11:46:43.814564+00
 p-15           |               1 | 2021-11-09 11:46:43.873654+00
 p-15           |               6 | 2021-11-09 11:46:43.894442+00

```

**with different inserts (no batch)***

```
persistence_id | sequence_number |         db_timestamp          
----------------+-----------------+-------------------------------
 p-1            |               1 | 2021-11-09 11:50:13.993737+00
 p-1            |               2 | 2021-11-09 11:50:13.993737+00
 p-1            |               3 | 2021-11-09 11:50:13.993737+00
 p-1            |               4 | 2021-11-09 11:50:13.993737+00
 p-1            |               5 | 2021-11-09 11:50:13.993737+00
 p-2            |               1 | 2021-11-09 11:50:14.200945+00
 p-2            |               2 | 2021-11-09 11:50:14.200945+00
 p-2            |               3 | 2021-11-09 11:50:14.200945+00
 p-2            |               4 | 2021-11-09 11:50:14.200945+00
 p-2            |               5 | 2021-11-09 11:50:14.200945+00
 p-3            |               1 | 2021-11-09 11:50:14.27671+00
 p-3            |               2 | 2021-11-09 11:50:14.27671+00
 p-3            |               3 | 2021-11-09 11:50:14.27671+00
 p-3            |               4 | 2021-11-09 11:50:14.27671+00
 p-3            |               5 | 2021-11-09 11:50:14.27671+00
 p-4            |               1 | 2021-11-09 11:50:14.348052+00
 p-4            |               2 | 2021-11-09 11:50:14.348052+00
 p-4            |               3 | 2021-11-09 11:50:14.348052+00
 p-4            |               4 | 2021-11-09 11:50:14.348052+00
 p-4            |               5 | 2021-11-09 11:50:14.348052+00
 p-5            |               1 | 2021-11-09 11:50:14.409302+00
 p-5            |               2 | 2021-11-09 11:50:14.409302+00
 p-5            |               3 | 2021-11-09 11:50:14.409302+00
 p-5            |               4 | 2021-11-09 11:50:14.409302+00
 p-5            |               5 | 2021-11-09 11:50:14.409302+00
 p-6            |               1 | 2021-11-09 11:50:14.471395+00
 p-6            |               2 | 2021-11-09 11:50:14.471395+00
 p-6            |               3 | 2021-11-09 11:50:14.471395+00
 p-6            |               4 | 2021-11-09 11:50:14.471395+00
 p-6            |               5 | 2021-11-09 11:50:14.471395+00
 p-7            |               1 | 2021-11-09 11:50:14.530211+00
 p-7            |               2 | 2021-11-09 11:50:14.530211+00
 p-7            |               3 | 2021-11-09 11:50:14.530211+00
 p-7            |               4 | 2021-11-09 11:50:14.530211+00
 p-7            |               5 | 2021-11-09 11:50:14.530211+00
 p-8            |               1 | 2021-11-09 11:50:14.585792+00
 p-8            |               2 | 2021-11-09 11:50:14.585792+00
 p-8            |               3 | 2021-11-09 11:50:14.585792+00
 p-8            |               4 | 2021-11-09 11:50:14.585792+00
 p-8            |               5 | 2021-11-09 11:50:14.585792+00
 p-9            |               1 | 2021-11-09 11:50:14.64328+00
 p-9            |               2 | 2021-11-09 11:50:14.64328+00
 p-9            |               3 | 2021-11-09 11:50:14.64328+00
 p-9            |               4 | 2021-11-09 11:50:14.64328+00
 p-9            |               5 | 2021-11-09 11:50:14.64328+00 
 p-10           |               1 | 2021-11-09 11:50:14.701976+00
 p-10           |               2 | 2021-11-09 11:50:14.701976+00
 p-10           |               3 | 2021-11-09 11:50:14.701976+00
 p-10           |               4 | 2021-11-09 11:50:14.701976+00
 p-10           |               5 | 2021-11-09 11:50:14.701976+00
 p-11           |               1 | 2021-11-09 11:50:14.754197+00
 p-11           |               2 | 2021-11-09 11:50:14.754197+00
 p-11           |               3 | 2021-11-09 11:50:14.754197+00
 p-11           |               4 | 2021-11-09 11:50:14.754197+00
 p-11           |               5 | 2021-11-09 11:50:14.754197+00
 p-12           |               4 | 2021-11-09 11:50:14.810084+00
 p-12           |               5 | 2021-11-09 11:50:14.810084+00
 p-12           |               3 | 2021-11-09 11:50:14.852788+00
 p-13           |               4 | 2021-11-09 11:50:15.107249+00
 p-13           |               5 | 2021-11-09 11:50:15.107249+00
 p-13           |               3 | 2021-11-09 11:50:15.143545+00
 p-15           |               1 | 2021-11-09 11:50:15.261206+00
 p-15           |               2 | 2021-11-09 11:50:15.261206+00
 p-15           |               3 | 2021-11-09 11:50:15.261206+00
 p-15           |               4 | 2021-11-09 11:50:15.261206+00
 p-15           |               5 | 2021-11-09 11:50:15.261206+00
 p-14           |               5 | 2021-11-09 11:50:15.217979+00
 p-15           |               6 | 2021-11-09 11:50:15.285976+00
```

This is the DB content after running `R2dbcJournalSpec`. 

Next, I run this test from the driver itself. 

https://github.com/pgjdbc/r2dbc-postgresql/blob/f2c2e484d434d29e32222c82afa215c3fcdbb35e/src/test/java/io/r2dbc/postgresql/PostgresCancelIntegrationTests.java#L85-L98

The test itself is verifying something else, but after `execute()` we should see two records in the DB, but instead we have:

```psql
test=# select * from insert_test;
 id | value
----+-------
  1 | a
```

**Comment by octonato** [comment]
One more piece of info, in the driver readme file it reads:

This driver provides the following features:
* Execution of prepared statements with bindings
* Execution of batch statements without bindings

This is not so clear though. Thera are Statements (which we are using) and Batch. Statement may have binding and Batch is indeed without bindings. 

So, I guess what is missing in the description is: 

* Execution of prepared statements with bindings in batch (not supported yet)

**Comment by patriknw** [comment]
Did you use it with `add()`, like:
```
connection.createStatement(...)
  .bind(..)
  .add()
  .bind(..)
  .execute()
```

**Comment by octonato** [comment]
Yep, exactly. That's also what the test in the driver does.

**Comment by octonato** [comment]
I talked with the maintainer on Gitter and we got the test in the driver in a state that works. Not clear why not working on our side. Digging....

**Comment by octonato** [comment]
I found the issue.

Although we execute a batch as one single statement, the returned Publisher will emit one element per insert. We were assuming that such a batch statement would return one single result. Our subscriber (asFuture) picks the first one and don't wait for the others. 

That doesn't feel much as a batch. It seems that it's just about constructing many inserts without recreating the statement each time. 

Like in...
```
INSERT INTO foo VALUES ("a");
INSERT INTO foo VALUES ("b");
```

Instead of...
```
INSERT INTO foo VALUES 
	("a"),
    ("b")
```

We need to consume the whole Publisher to get all the inserts. I got something working, now just need to do some clean-up. I got it a little messy.
